optaplanner-examples/src/main/java/org/optaplanner/examples/common/experimental/impl/SequenceImpl.java

```diff
@@ -56,62 +58,62 @@ public ValueType_ getLastItem() {
 
     @Override
     public NavigableSet<ValueType_> getItems() {
-        return consecutiveItemsSet;
+        if (items == null) {
+            return items = sourceTree.getItemSet()
+                    .subSet(firstItem, true, lastItem, true);
+        }
+        return items;
     }
 
     @Override
     public int getCount() {
-        return consecutiveItemsSet.size();
+        return getItems().size();
     }
 
     @Override
     public DifferenceType_ getLength() {
-        return sourceTree.getSequenceLength(this);
-    }
-
-    protected boolean isEmpty() {
-        return consecutiveItemsSet.isEmpty();
+        if (length == null) {
+            // memoize length for later calls
+            // (assignment returns the right hand side)
+            return length = sourceTree.getSequenceLength(this);
+        }
+        return length;
     }
 
-    protected void addBeforeStart(ValueType_ item) {
+    protected void setStart(ValueType_ item) {
         firstItem = item;
-        consecutiveItemsSet = sourceTree.getItemSet().subSet(item, true, lastItem, true);
+        invalidate();
     }
 
-    protected void addAfterEnd(ValueType_ item) {
+    protected void setEnd(ValueType_ item) {
         lastItem = item;
-        consecutiveItemsSet = sourceTree.getItemSet().subSet(firstItem, true, item, true);
+        invalidate();
     }
 
-    // Called when start or end are removed; since the set
-    // is backed by the underlying set that already have the
-    // point removed, consecutiveItemsSet.first()/consecutiveItemsSet.last()
-    // change to reflect that, and we update the bounds
-    protected void updateEndpoints() {
-        firstItem = consecutiveItemsSet.first();
-        lastItem = consecutiveItemsSet.last();
-        consecutiveItemsSet = consecutiveItemsSet.subSet(firstItem, true, lastItem, true);
+    // Called when start or end are removed; length
+    // need to be invalidated
+    protected void invalidate() {
+        length = null;
+        items = null;
     }
```

Comment:
We still have this problem that changing state of the sequence requires two operations.
Shouldn't we be able to invalidate automatically when start/end changes?

Reply:
set start and set end do call invalidate; I think I need to remove this extra call from remove